In [60]:
#!/usr/bin/env python
# coding: utf-8

#### Load the Data

In[1]:

In [61]:
import pandas as pd
import numpy as np
import re
import warnings
warnings.filterwarnings('ignore')
import seaborn as sns
#from wordcloud import WordCloud, STOPWORDS, ImageColorGenerator
import matplotlib.cm as cm
import matplotlib as plt
get_ipython().run_line_magic('matplotlib', 'inline')

In[2]:

oad the full set of the data<br>
Read excel file

In [62]:
df = pd.read_excel('input_data.xlsx')
# Shape and size of the dataframe
print("> Shape of input_data :",df.shape)
print("\n> Size of input_data :",df.size)

> Shape of input_data : (8500, 4)

> Size of input_data : 34000


In[3]:

heck if the data frame is properly loaded using the sample() method

In [63]:
df.sample(10)

Short description  \
1545  job Job_593 failed in job_scheduler at: 10/13/...   
7805                           skype funktionert nicht.   
4184               install pdfmailer \ahlqgjwx wbsfavhg   
4010  job Job_459 failed in job_scheduler at: 09/19/...   
7761  network problems (multiple applications are ru...   
2223  job Job_3050 failed in job_scheduler at: 10/05...   
1810                     carrier orders not processing.   
6992             probleme mit erpgui \tmqfjard qzhgdoua   
6884  HostName_1132:oracle finance_app financial man...   
1059  netzteil oder netzstecker defekt - pc warenein...   

                                            Description             Caller  \
1545  received from: monitoring_tool@company.com_x00...  bpctwhsn kzqsbmtp   
7805  from: zlnxswvp ptmzsbhk _x000D_\nsent: thursda...  zlnxswvp ptmzsbhk   
4184               install pdfmailer \ahlqgjwx wbsfavhg  ahlqgjwx wbsfavhg   
4010  received from: monitoring_tool@company.com_x00...  bpctwhsn kzqsbmtp   
7761  how did you determine there are network proble...  gbxctqfl akrwmuib   
2223  received from: monitoring_tool@company.com_x00...  bpctwhsn kzqsbmtp   
1810  carrier orders not processing labels through s...  tmyeqika hfudpeot   
6992             probleme mit erpgui \tmqfjard qzhgdoua  tmqfjard qzhgdoua   
6884  HostName_1132:oracle finance_app financial man...  oldrctiu bxurpsyi   
1059  bitte netzteil oder netzstecker am pc evhw8114...  ipeajlvk idmbyztf   

     Assignment group  
1545            GRP_8  
7805            GRP_0  
4184           GRP_24  
4010            GRP_8  
7761            GRP_0  
2223            GRP_8  
1810           GRP_18  
6992           GRP_24  
6884           GRP_12  
1059           GRP_33

In[4]:

heck if the data frame is properly loaded using the head() method

In [64]:
df.head(10)

Short description  \
0                                        login issue   
1                                            outlook   
2                                 cant log in to vpn   
3                      unable to access hr_tool page   
4                                       skype error    
5     unable to log in to engineering tool and skype   
6  event: critical:HostName_221.company.com the v...   
7  ticket_no1550391- employment status - new non-...   
8               unable to disable add ins on outlook   
9                    ticket update on inplant_874773   

                                         Description             Caller  \
0  -verified user details.(employee# & manager na...  spxjnwir pjlcoqds   
1  _x000D_\n_x000D_\nreceived from: hmjdrvpb.komu...  hmjdrvpb komuaywn   
2  _x000D_\n_x000D_\nreceived from: eylqgodm.ybqk...  eylqgodm ybqkwiam   
3                      unable to access hr_tool page  xbkucsvz gcpydteq   
4                                       skype error   owlgqjme qhcozdfx   
5     unable to log in to engineering tool and skype  eflahbxn ltdgrvkz   
6  event: critical:HostName_221.company.com the v...  jyoqwxhz clhxsoqy   
7  ticket_no1550391- employment status - new non-...  eqzibjhw ymebpoih   
8               unable to disable add ins on outlook  mdbegvct dbvichlg   
9                    ticket update on inplant_874773  fumkcsji sarmtlhy   

  Assignment group  
0            GRP_0  
1            GRP_0  
2            GRP_0  
3            GRP_0  
4            GRP_0  
5            GRP_0  
6            GRP_1  
7            GRP_0  
8            GRP_0  
9            GRP_0

In[5]:

heck if the data frame is properly loaded using the tail() method

In [65]:
df.tail(10)

Short description  \
8490                         check status in purchasing   
8491                                     vpn for laptop   
8492                 hr_tool etime option not visitble    
8493           erp fi  - ob09, two accounts to be added   
8494  tablet needs reimaged due to multiple issues w...   
8495                  emails not coming in from zz mail   
8496                           telephony_software issue   
8497  vip2: windows password reset for tifpdchb pedx...   
8498                     machine não está funcionando   
8499  an mehreren pc`s lassen sich verschiedene prgr...   

                                            Description             Caller  \
8490  please contact ed pasgryowski (pasgryo) about ...  mpihysnw wrctgoan   
8491  \n\nreceived from: jxgobwrm.qkugdipo@gmail.com...  jxgobwrm qkugdipo   
8492                 hr_tool etime option not visitble   tmopbken ibzougsd   
8493  i am sorry, i have another two accounts that n...  ipwjorsc uboapexr   
8494  tablet needs reimaged due to multiple issues w...  cpmaidhj elbaqmtp   
8495  _x000D_\n_x000D_\nreceived from: avglmrts.vhqm...  avglmrts vhqmtiua   
8496                           telephony_software issue  rbozivdq gmlhrtvp   
8497  vip2: windows password reset for tifpdchb pedx...  oybwdsgx oxyhwrfz   
8498  i am unable to access the machine utilities to...  ufawcgob aowhxjky   
8499  an mehreren pc`s lassen sich verschiedene prgr...  kqvbrspl jyzoklfx   

     Assignment group  
8490           GRP_29  
8491           GRP_34  
8492            GRP_0  
8493           GRP_10  
8494            GRP_3  
8495           GRP_29  
8496            GRP_0  
8497            GRP_0  
8498           GRP_62  
8499           GRP_49

In[6]:

In [66]:
df.columns

Index(['Short description', 'Description', 'Caller', 'Assignment group'], dtype='object')

In[7]:

heck the feature/columns  using the info method

In [67]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8500 entries, 0 to 8499
Data columns (total 4 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Short description  8492 non-null   object
 1   Description        8499 non-null   object
 2   Caller             8500 non-null   object
 3   Assignment group   8500 non-null   object
dtypes: object(4)
memory usage: 265.8+ KB


In[8]:

In [68]:
df['Assignment group'].unique()

array(['GRP_0', 'GRP_1', 'GRP_3', 'GRP_4', 'GRP_5', 'GRP_6', 'GRP_7',
       'GRP_8', 'GRP_9', 'GRP_10', 'GRP_11', 'GRP_12', 'GRP_13', 'GRP_14',
       'GRP_15', 'GRP_16', 'GRP_17', 'GRP_18', 'GRP_19', 'GRP_2',
       'GRP_20', 'GRP_21', 'GRP_22', 'GRP_23', 'GRP_24', 'GRP_25',
       'GRP_26', 'GRP_27', 'GRP_28', 'GRP_29', 'GRP_30', 'GRP_31',
       'GRP_33', 'GRP_34', 'GRP_35', 'GRP_36', 'GRP_37', 'GRP_38',
       'GRP_39', 'GRP_40', 'GRP_41', 'GRP_42', 'GRP_43', 'GRP_44',
       'GRP_45', 'GRP_46', 'GRP_47', 'GRP_48', 'GRP_49', 'GRP_50',
       'GRP_51', 'GRP_52', 'GRP_53', 'GRP_54', 'GRP_55', 'GRP_56',
       'GRP_57', 'GRP_58', 'GRP_59', 'GRP_60', 'GRP_61', 'GRP_32',
       'GRP_62', 'GRP_63', 'GRP_64', 'GRP_65', 'GRP_66', 'GRP_67',
       'GRP_68', 'GRP_69', 'GRP_70', 'GRP_71', 'GRP_72', 'GRP_73'],
      dtype=object)

In[9]:

In [69]:
df['Assignment group'].nunique()

74

NOTE : There are 74 Dependent/Target classes which are required to be predicted by our Model

In[10]:

In [70]:
targetClassCnt=df['Assignment group'].value_counts()
targetClassCnt

GRP_0     3976
GRP_8      661
GRP_24     289
GRP_12     257
GRP_9      252
          ... 
GRP_64       1
GRP_67       1
GRP_35       1
GRP_70       1
GRP_73       1
Name: Assignment group, Length: 74, dtype: int64

In[11]:

In [71]:
targetClassCnt.describe()

count      74.000000
mean      114.864865
std       465.747516
min         1.000000
25%         5.250000
50%        26.000000
75%        84.000000
max      3976.000000
Name: Assignment group, dtype: float64

#### Observation:<br>
- Appears the Target class distrubtion is extremely skewed<br>
- Large no of entries for GRP_0 (mounting to 3976) which account for ~50% of the data<br>
- there are groups with 1 entry also. We could merge all groups with small entries to a group to reduce the imbalance in the target. This may reduce the imbalance to some extent.

In[12]:

hceck for na values

In [72]:
df.isna().sum()

Short description    8
Description          1
Caller               0
Assignment group     0
dtype: int64

NOTE : There are 8 NaN Values in 'Short description' Feature, and 1 NaN value in 'Description' Feature.

In[13]:

 find nulls

In [73]:
df[df.isnull().any(axis=1)]

Short description  \
2604                       NaN   
3383                       NaN   
3906                       NaN   
3910                       NaN   
3915                       NaN   
3921                       NaN   
3924                       NaN   
4341                       NaN   
4395  i am locked out of skype   

                                            Description             Caller  \
2604  _x000D_\n_x000D_\nreceived from: ohdrnswl.rezu...  ohdrnswl rezuibdt   
3383  _x000D_\n-connected to the user system using t...  qftpazns fxpnytmk   
3906  -user unable  tologin to vpn._x000D_\n-connect...  awpcmsey ctdiuqwe   
3910  -user unable  tologin to vpn._x000D_\n-connect...  rhwsmefo tvphyura   
3915  -user unable  tologin to vpn._x000D_\n-connect...  hxripljo efzounig   
3921  -user unable  tologin to vpn._x000D_\n-connect...  cziadygo veiosxby   
3924  name:wvqgbdhm fwchqjor\nlanguage:\nbrowser:mic...  wvqgbdhm fwchqjor   
4341  _x000D_\n_x000D_\nreceived from: eqmuniov.ehxk...  eqmuniov ehxkcbgj   
4395                                                NaN  viyglzfo ajtfzpkb   

     Assignment group  
2604           GRP_34  
3383            GRP_0  
3906            GRP_0  
3910            GRP_0  
3915            GRP_0  
3921            GRP_0  
3924            GRP_0  
4341            GRP_0  
4395            GRP_0

In[14]:

For Short Description

In [74]:
df[df['Short description'].isnull()]

Short description                                        Description  \
2604               NaN  _x000D_\n_x000D_\nreceived from: ohdrnswl.rezu...   
3383               NaN  _x000D_\n-connected to the user system using t...   
3906               NaN  -user unable  tologin to vpn._x000D_\n-connect...   
3910               NaN  -user unable  tologin to vpn._x000D_\n-connect...   
3915               NaN  -user unable  tologin to vpn._x000D_\n-connect...   
3921               NaN  -user unable  tologin to vpn._x000D_\n-connect...   
3924               NaN  name:wvqgbdhm fwchqjor\nlanguage:\nbrowser:mic...   
4341               NaN  _x000D_\n_x000D_\nreceived from: eqmuniov.ehxk...   

                 Caller Assignment group  
2604  ohdrnswl rezuibdt           GRP_34  
3383  qftpazns fxpnytmk            GRP_0  
3906  awpcmsey ctdiuqwe            GRP_0  
3910  rhwsmefo tvphyura            GRP_0  
3915  hxripljo efzounig            GRP_0  
3921  cziadygo veiosxby            GRP_0  
3924  wvqgbdhm fwchqjor            GRP_0  
4341  eqmuniov ehxkcbgj            GRP_0

In[15]:

For Description

In [75]:
df[df['Description'].isnull()]

Short description Description             Caller Assignment group
4395  i am locked out of skype         NaN  viyglzfo ajtfzpkb            GRP_0

In[16]:

NULL replacement<br>
replacing nan values in short description

In [76]:
df["Short description"].fillna(df["Description"], inplace = True)

In[17]:

replacing nan values in description

In [77]:
df["Description"].fillna(df["Short description"], inplace = True)

In[18]:

verify the replacement

In [78]:
df.isnull().sum()

Short description    0
Description          0
Caller               0
Assignment group     0
dtype: int64

In[19]:

verify the replacement

In [79]:
print((df.iloc[2604]))

Short description    _x000D_\n_x000D_\nreceived from: ohdrnswl.rezu...
Description          _x000D_\n_x000D_\nreceived from: ohdrnswl.rezu...
Caller                                               ohdrnswl rezuibdt
Assignment group                                                GRP_34
Name: 2604, dtype: object


In[20]:

verify the replacement

In [80]:
print((df.iloc[4395]))

Short description    i am locked out of skype
Description          i am locked out of skype
Caller                      viyglzfo ajtfzpkb
Assignment group                        GRP_0
Name: 4395, dtype: object


In[21]:

dentify duplicate rows

In [81]:
duplicateRows = df[df.duplicated()]

In[22]:

iew duplicate rows

In [82]:
duplicateRows

Short description  \
51                           call for ecwtrjnq jpecxuty   
229                          call for ecwtrjnq jpecxuty   
493                     ticket update on inplant_872730   
512                                    blank call //gso   
667   job bkbackup_tool_powder_prod_full failed in j...   
...                                                 ...   
7836             probleme mit erpgui \tmqfjard qzhgdoua   
8051               issue on pricing in distributor_tool   
8093  reset passwords for prgthyuulla ramdntythanjes...   
8347                           blank call // loud noise   
8405                           unable to launch outlook   

                                            Description             Caller  \
51                           call for ecwtrjnq jpecxuty  olckhmvx pcqobjnd   
229                          call for ecwtrjnq jpecxuty  olckhmvx pcqobjnd   
493                     ticket update on inplant_872730  fumkcsji sarmtlhy   
512                                    blank call //gso  rbozivdq gmlhrtvp   
667   received from: monitoring_tool@company.com_x00...  bpctwhsn kzqsbmtp   
...                                                 ...                ...   
7836             probleme mit erpgui \tmqfjard qzhgdoua  tmqfjard qzhgdoua   
8051  we have agreed price with many of the distribu...  hbmwlprq ilfvyodx   
8093                                                the  boirqctx bkijgqry   
8347                           blank call // loud noise  rbozivdq gmlhrtvp   
8405                           unable to launch outlook  wjtzrmqc ikqpbflg   

     Assignment group  
51              GRP_0  
229             GRP_0  
493             GRP_0  
512             GRP_0  
667             GRP_8  
...               ...  
7836           GRP_24  
8051           GRP_21  
8093           GRP_17  
8347            GRP_0  
8405            GRP_0  

[83 rows x 4 columns]

In[23]:

drop duplicate rows

In [83]:
df.drop_duplicates()

Short description  \
0                                           login issue   
1                                               outlook   
2                                    cant log in to vpn   
3                         unable to access hr_tool page   
4                                          skype error    
...                                                 ...   
8495                  emails not coming in from zz mail   
8496                           telephony_software issue   
8497  vip2: windows password reset for tifpdchb pedx...   
8498                     machine não está funcionando   
8499  an mehreren pc`s lassen sich verschiedene prgr...   

                                            Description             Caller  \
0     -verified user details.(employee# & manager na...  spxjnwir pjlcoqds   
1     _x000D_\n_x000D_\nreceived from: hmjdrvpb.komu...  hmjdrvpb komuaywn   
2     _x000D_\n_x000D_\nreceived from: eylqgodm.ybqk...  eylqgodm ybqkwiam   
3                         unable to access hr_tool page  xbkucsvz gcpydteq   
4                                          skype error   owlgqjme qhcozdfx   
...                                                 ...                ...   
8495  _x000D_\n_x000D_\nreceived from: avglmrts.vhqm...  avglmrts vhqmtiua   
8496                           telephony_software issue  rbozivdq gmlhrtvp   
8497  vip2: windows password reset for tifpdchb pedx...  oybwdsgx oxyhwrfz   
8498  i am unable to access the machine utilities to...  ufawcgob aowhxjky   
8499  an mehreren pc`s lassen sich verschiedene prgr...  kqvbrspl jyzoklfx   

     Assignment group  
0               GRP_0  
1               GRP_0  
2               GRP_0  
3               GRP_0  
4               GRP_0  
...               ...  
8495           GRP_29  
8496            GRP_0  
8497            GRP_0  
8498           GRP_62  
8499           GRP_49  

[8417 rows x 4 columns]

#### Data Cleaning and Preprocessing

#### Now clean up the Data to address the following<br>
<br>
- Convert each character in a sentence to lowercase character<br>
- Remove HTML Tags<br>
- Remove punctuations<br>
- Remove stopwords<br>
- Remove common words like com, hello<br>
- Remove NUmbers - (r'[0-9] , is used for replace, but there are some numbers which remain in the data.. that needs to be tested further<br>
- Stemming was causing invalid words, hence used a lemmatizer

In[24]:

In [84]:
import nltk
from nltk.tokenize import RegexpTokenizer
from nltk.stem import WordNetLemmatizer,PorterStemmer
from nltk.corpus import stopwords
import re
lemmatizer = WordNetLemmatizer()
stemmer = PorterStemmer() 

In [85]:
def preprocess(sentence):
    sentence=str(sentence)
    sentence = sentence.lower()
    sentence=sentence.replace('{html}',"") 
    cleanr = re.compile('<.*?>')
    cleantext = re.sub(cleanr, '', sentence)
    rem_url=re.sub(r'http\S+', '',cleantext)
    rem_num = re.sub('[0-9]+', '', rem_url)
    rem = re.sub('_xd_', '', rem_num)
    rem1 = re.sub('_', '', rem)
    tokenizer = RegexpTokenizer(r'\w+')
    tokens = tokenizer.tokenize(rem1)  
    filtered_words = [w for w in tokens if len(w) > 2 if not w in stopwords.words('english')]
    stem_words=[stemmer.stem(w) for w in filtered_words]
    lemma_words=[lemmatizer.lemmatize(w) for w in stem_words]
    clean_words = " ".join(filtered_words)
    return clean_words

In[25]:

In [86]:
import nltk
nltk.download('stopwords')
nltk.download('wordnet')

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\KumarAbh5\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\KumarAbh5\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


True

In [87]:
df['Short description']=df['Short description'].map(lambda s:preprocess(s)) 

In[26]:

In [88]:
df['Description']=df['Description'].map(lambda s:preprocess(s)) 

In[27]:

In [89]:
df['Caller']=df['Caller'].map(lambda s:preprocess(s)) 

In[28]:

In [90]:
df.head(10)

Short description  \
0                                        login issue   
1                                            outlook   
2                                       cant log vpn   
3                          unable access hrtool page   
4                                        skype error   
5                  unable log engineering tool skype   
6  event critical hostname company com value moun...   
7  ticketno employment status new non employee en...   
8                     unable disable add ins outlook   
9                              ticket update inplant   

                                         Description             Caller  \
0  verified user details employee manager name ch...  spxjnwir pjlcoqds   
1  received hmjdrvpb komuaywn gmail com hello tea...  hmjdrvpb komuaywn   
2  received eylqgodm ybqkwiam gmail com cannot lo...  eylqgodm ybqkwiam   
3                          unable access hrtool page  xbkucsvz gcpydteq   
4                                        skype error  owlgqjme qhcozdfx   
5                  unable log engineering tool skype  eflahbxn ltdgrvkz   
6  event critical hostname company com value moun...  jyoqwxhz clhxsoqy   
7  ticketno employment status new non employee en...  eqzibjhw ymebpoih   
8                     unable disable add ins outlook  mdbegvct dbvichlg   
9                              ticket update inplant  fumkcsji sarmtlhy   

  Assignment group  
0            GRP_0  
1            GRP_0  
2            GRP_0  
3            GRP_0  
4            GRP_0  
5            GRP_0  
6            GRP_1  
7            GRP_0  
8            GRP_0  
9            GRP_0

In[29]:

In [91]:
df.tail(10)

Short description  \
8490                            check status purchasing   
8491                                         vpn laptop   
8492                       hrtool etime option visitble   
8493                             erp two accounts added   
8494  tablet needs reimaged due multiple issues crm ...   
8495                                 emails coming mail   
8496                            telephonysoftware issue   
8497       vip windows password reset tifpdchb pedxruyf   
8498                           machine estã funcionando   
8499  mehreren lassen sich verschiedene prgramdntyme...   

                                            Description             Caller  \
8490  please contact pasgryowski pasgryo purchasing ...  mpihysnw wrctgoan   
8491  received jxgobwrm qkugdipo gmail com need vpn ...  jxgobwrm qkugdipo   
8492                       hrtool etime option visitble  tmopbken ibzougsd   
8493  sorry another two accounts need added please c...  ipwjorsc uboapexr   
8494  tablet needs reimaged due multiple issues crm ...  cpmaidhj elbaqmtp   
8495  received avglmrts vhqmtiua gmail com good afte...  avglmrts vhqmtiua   
8496                            telephonysoftware issue  rbozivdq gmlhrtvp   
8497       vip windows password reset tifpdchb pedxruyf  oybwdsgx oxyhwrfz   
8498  unable access machine utilities finish drawers...  ufawcgob aowhxjky   
8499  mehreren lassen sich verschiedene prgramdntyme...  kqvbrspl jyzoklfx   

     Assignment group  
8490           GRP_29  
8491           GRP_34  
8492            GRP_0  
8493           GRP_10  
8494            GRP_3  
8495           GRP_29  
8496            GRP_0  
8497            GRP_0  
8498           GRP_62  
8499           GRP_49

In[30]:

In [92]:
df.sample(10)

Short description  \
5237             probleme mit outlook axesnghb cyzuomxa   
4281        able apply job company portal external user   
3480  bill hartwell access drawings hostname departm...   
539   cracked screen turned touch screen longer jump...   
8411                 password reset supplychainsoftware   
967                                            passwort   
3736                        job job failed jobscheduler   
6882                        job job failed jobscheduler   
5508            nmpworvu upgtrvnj monitor display issue   
3286                                truview application   

                                            Description             Caller  \
5237             probleme mit outlook axesnghb cyzuomxa  axesnghb cyzuomxa   
4281        able apply job company portal external user  efbwiadp dicafxhv   
3480  please assign bill necessary permissions open ...  zqrnveyx hajtfmcd   
539   cracked screen turned touch screen longer jump...  jkelxorm uwqjlzfm   
8411  password reset supplychainsoftware user ludwid...  xeucniqa dwpsevof   
967   received soldfnbq uhnbsvqd gmail com hallo ers...  soldfnbq uhnbsvqd   
3736  received monitoringtool company com job job fa...  bpctwhsn kzqsbmtp   
6882  received monitoringtool company com job job fa...  bpctwhsn kzqsbmtp   
5508            nmpworvu upgtrvnj monitor display issue  rbozivdq gmlhrtvp   
3286  truview application getting alarms also saw to...  wczegmok bgqoclvs   

     Assignment group  
5237           GRP_24  
4281            GRP_0  
3480           GRP_12  
539             GRP_3  
8411            GRP_0  
967             GRP_0  
3736            GRP_8  
6882            GRP_8  
5508            GRP_0  
3286            GRP_4

#### Observations:<br>
Index 2865 has sentences in different language than English. So we need to perform Translation for such rows.

In[31]:

In [93]:
get_ipython().system('pip install deep-translator')

In[69]:

ranslation = translator.translate("¾ç½ ¹äº ï¼œæ žä¹ˆè šåž")<br>
rom deep_translator import GoogleTranslator<br>
ranslation = GoogleTranslator(source='auto', target='en').translate("maus funktioniert nicht mehr richtig bitte")<br>
ranslation

In[121]:

In [94]:
germanwordlist = ['bitte','nicht','konto','probleme','berechtigung','defekt','mehr','ausgetauscht','rechner', 'drucker','teilweise','freigegeben','genannten','anmeldeaccount',
                  'besprochen','werden','durchwahl','oben','einrichten','zeitwirtschaft','seit','morgens','beheben','keine','zeitbuchungen','vorhanden','dringend','fehler',
                  'werk','anmelde','auftrag','kein','skannen','freundlichen','werkzeuge,','hartstoffe','maste','schutzw','fertigung','immer','sehr','zugriff','freundliche',
                  'geehrter','souzarft','noch','verbindungsherstellung','meldung','erneuten','glich','proben','beilageproben','beilage','auswerten','sinterleitstand','reparar',
                  'reparo','rechner','koenigsee','entregar','atualiza','declara','programdntya','funcionando','preciso','hitacni','grergtger','zugriffsrechte','teamleiter',
                  'abholen','wegen','weit','absender','wenn','abrechnung']
pattern = '|'.join(germanwordlist)

pattern

'bitte|nicht|konto|probleme|berechtigung|defekt|mehr|ausgetauscht|rechner|drucker|teilweise|freigegeben|genannten|anmeldeaccount|besprochen|werden|durchwahl|oben|einrichten|zeitwirtschaft|seit|morgens|beheben|keine|zeitbuchungen|vorhanden|dringend|fehler|werk|anmelde|auftrag|kein|skannen|freundlichen|werkzeuge,|hartstoffe|maste|schutzw|fertigung|immer|sehr|zugriff|freundliche|geehrter|souzarft|noch|verbindungsherstellung|meldung|erneuten|glich|proben|beilageproben|beilage|auswerten|sinterleitstand|reparar|reparo|rechner|koenigsee|entregar|atualiza|declara|programdntya|funcionando|preciso|hitacni|grergtger|zugriffsrechte|teamleiter|abholen|wegen|weit|absender|wenn|abrechnung'

In[ ]:

In [95]:
germanDescIndex = df[df['Description'].str.contains(pattern)].index
germanDescIndex

Int64Index([ 223,  238,  251,  255,  265,  270,  272,  302,  304,  305,
            ...
            8414, 8417, 8419, 8420, 8425, 8427, 8436, 8439, 8467, 8499],
           dtype='int64', length=589)

In [99]:
!pip install -q googletrans
from googletrans import Translator

def translateIfRequired(x):
    translator = Translator()
    if translator.detect(x).lang != 'en':
        translatedText = translator.translate(x).text
    else:
        translatedText = x
    return translatedText

In [98]:
for rowId in germanDescIndex:
    df['Description'][rowId] = translateIfRequired(df['Description'][rowId])

AttributeError: 'NoneType' object has no attribute 'group'

In [100]:
germanDescIndex_new = df[df['Description'].str.contains(pattern)].index
germanDescIndex_new

Int64Index([ 223,  238,  251,  255,  265,  270,  272,  302,  304,  305,
            ...
            8414, 8417, 8419, 8420, 8425, 8427, 8436, 8439, 8467, 8499],
           dtype='int64', length=589)

In [101]:
df['Description'][8232]

'stepfhryhan needs access collaborationplatform links urlaubsplanung file efdl users linnes collaborationplatform company inc teamordner linnemann urlaubsplanung plã file efdl users linnes collaborationplatform company inc teamordner linnemann ene allgemeines file efdl users linnes collaborationplatform company inc teamordner linnemann allgemeines berirtchtswesen gebiet nord file efdl users linnes collaborationplatform company inc teamordner linnemann berirtchtswesen gebiet nord crm teamordner file efdl users linnes collaborationplatform company inc teamordner linnemann crm teamordner teamcall teammeeting file efdl users linnes collaborationplatform company inc teamordner linnemann teamcall teammeeting top projekte file efdl users linnes collaborationplatform company inc teamordner linnemann top projekte hallo sabrthy wie weit sind wir diesem thema habe noch kein ticket erhaltenâ mit freundlichen grã¼ãÿen nizholae bjnqikym anwendungstechniker application engineer nizholae bjnqikym gmai

In [ ]:
df['Short description'] = df['Short description'].apply(lambda x: GoogleTranslator(source='auto', target='en').translate(str(x)))

In[42]:

verify the replacement

In [ ]:
print((df.iloc[3983]))

#### Visualize the Dsitribution of Records across Groups

In[32]:

In [102]:
import string
import collections

In[33]:

In [103]:
all_data = df['Description'].str.split(' ')
all_data.head()

0    [verified, user, details, employee, manager, n...
1    [received, hmjdrvpb, komuaywn, gmail, com, hel...
2    [received, eylqgodm, ybqkwiam, gmail, com, can...
3                       [unable, access, hrtool, page]
4                                       [skype, error]
Name: Description, dtype: object

In[34]:

In [104]:
all_data_cleaned = []

In [105]:
for text in all_data:
    text = [x.strip(string.punctuation) for x in text]
    all_data_cleaned.append(text)

In [106]:
all_data_cleaned

[['verified',
  'user',
  'details',
  'employee',
  'manager',
  'name',
  'checked',
  'user',
  'name',
  'reset',
  'password',
  'advised',
  'user',
  'login',
  'check',
  'caller',
  'confirmed',
  'able',
  'login',
  'issue',
  'resolved'],
 ['received',
  'hmjdrvpb',
  'komuaywn',
  'gmail',
  'com',
  'hello',
  'team',
  'meetings',
  'skype',
  'meetings',
  'etc',
  'appearing',
  'outlook',
  'calendar',
  'somebody',
  'please',
  'advise',
  'correct',
  'kind'],
 ['received',
  'eylqgodm',
  'ybqkwiam',
  'gmail',
  'com',
  'cannot',
  'log',
  'vpn',
  'best'],
 ['unable', 'access', 'hrtool', 'page'],
 ['skype', 'error'],
 ['unable', 'log', 'engineering', 'tool', 'skype'],
 ['event',
  'critical',
  'hostname',
  'company',
  'com',
  'value',
  'mountpoint',
  'threshold',
  'oracle',
  'sid',
  'erpdata',
  'srpsad',
  'srpsad',
  'data',
  'perpsrpsad'],
 ['ticketno',
  'employment',
  'status',
  'new',
  'non',
  'employee',
  'enter',
  'user',
  'name'],
 ['

In [44]:
all_data_cleaned[0]

['verified',
 'user',
 'details',
 'employee',
 'manager',
 'name',
 'checked',
 'user',
 'name',
 'reset',
 'password',
 'advised',
 'user',
 'login',
 'check',
 'caller',
 'confirmed',
 'able',
 'login',
 'issue',
 'resolved']

In [45]:
text_data = [" ".join(text) for text in all_data_cleaned]
final_text_data = " ".join(text_data)
final_text_data[:500]

'verified user details employee manager name checked user name reset password advised user login check caller confirmed able login issue resolved received hmjdrvpb komuaywn gmail com hello team meetings skype meetings etc appearing outlook calendar somebody please advise correct kind received eylqgodm ybqkwiam gmail com cannot log vpn best unable access hrtool page skype error unable log engineering tool skype event critical hostname company com value mountpoint threshold oracle sid erpdata srpsa'

In[35]:

In [48]:
pip install wordcloud

  Using cached wordcloud-1.8.1.tar.gz (220 kB)
  Running setup.py clean for wordcloud
Failed to build wordcloud
    Running setup.py install for wordcloud: started
    Running setup.py install for wordcloud: finished with status 'error'
Note: you may need to restart the kernel to use updated packages.


  ERROR: Command errored out with exit status 1:
   command: 'C:\Users\KumarAbh5\Anaconda3\python.exe' -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'C:\\Users\\KumarAbh5\\AppData\\Local\\Temp\\pip-install-aqmzrjqs\\wordcloud_010778dea9514442be354c34e85cd33c\\setup.py'"'"'; __file__='"'"'C:\\Users\\KumarAbh5\\AppData\\Local\\Temp\\pip-install-aqmzrjqs\\wordcloud_010778dea9514442be354c34e85cd33c\\setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d 'C:\Users\KumarAbh5\AppData\Local\Temp\pip-wheel-g33hlyre'
       cwd: C:\Users\KumarAbh5\AppData\Local\Temp\pip-install-aqmzrjqs\wordcloud_010778dea9514442be354c34e85cd33c\
  Complete output (20 lines):
  running bdist_wheel
  running build
  running build_py
  creating build
  creating bui

In [49]:
from wordcloud import WordCloud 

ModuleNotFoundError: No module named 'wordcloud'

In [46]:
import matplotlib.pyplot as pplt
wordcloud_desc = WordCloud(background_color="white").generate(final_text_data)

NameError: name 'WordCloud' is not defined

In [ ]:
pplt.figure(figsize = (20,20))
pplt.imshow(wordcloud_desc, interpolation='bilinear')
pplt.axis("off")
pplt.show()

In[36]:

In [ ]:
all_data1 = df['Short description'].str.split(' ')
all_data1.head()

In[37]:

In [ ]:
all_data_cleaned1 = []

In [ ]:
for text in all_data1:
    text = [x.strip(string.punctuation) for x in text]
    all_data_cleaned1.append(text)

In [ ]:
all_data_cleaned1[0]

In [ ]:
text_data1 = [" ".join(text) for text in all_data_cleaned1]
final_text_data1 = " ".join(text_data1)
final_text_data1[:500]

In[38]:

In [ ]:
wordcloud_sdesc = WordCloud(background_color="white").generate(final_text_data1)

In [ ]:
pplt.figure(figsize = (20,20))
pplt.imshow(wordcloud_sdesc, interpolation='bilinear')
pplt.axis("off")
pplt.show()

In[39]:

In [ ]:
all_data2 = df['Caller'].str.split(' ')
all_data2.head()

In[40]:

In [ ]:
all_data_cleaned2 = []

In [ ]:
for text in all_data2:
    text = [x.strip(string.punctuation) for x in text]
    all_data_cleaned2.append(text)

In [ ]:
all_data_cleaned2[0]

In [ ]:
text_data2 = [" ".join(text) for text in all_data_cleaned2]
final_text_data2 = " ".join(text_data2)
final_text_data2[:500]

In[41]:

In [ ]:
wordcloud_caller = WordCloud(background_color="white").generate(final_text_data2)

In [ ]:
pplt.figure(figsize = (20,20))
pplt.imshow(wordcloud_caller, interpolation='bilinear')
pplt.axis("off")
pplt.show()

In[58]:

 create a column to mark records with GRP_0 and non GRP_0=>GRP_X

In [ ]:
df['GRP_MOD'] = df['Assignment group'].apply(lambda x: 'GRP_X' if x != 'GRP_0' else x)

In[59]:

In [ ]:
stopwords = set(STOPWORDS)
## function to create Word Cloud
def show_wordcloud(data, title):
    wordcloud = WordCloud(
        background_color='white',
        stopwords=stopwords,
        max_words=200,
        max_font_size=40, 
        scale=3,
        random_state=1 # chosen at random by flipping a coin; it was heads
    ).generate(str(data))
    fig = plt.figure(1, figsize=(15, 12))
    plt.axis('off')
    if title: 
        fig.suptitle(title, fontsize=20)
        fig.subplots_adjust(top=2.3)
    plt.imshow(wordcloud)
    plt.show()

In[60]:

 view word cloud for GRP_0

In [ ]:
text_Str = df['Description'][df['GRP_MOD'].isin(["GRP_0"])].tolist()
show_wordcloud(text_Str, "GRP_0 WORD CLOUD")

In[61]:

 GRP_X Visualization:

In [ ]:
text_Str = df['Description'][df['GRP_MOD'].isin(["GRP_X"])].tolist()
show_wordcloud(text_Str, "GRP_X WORD CLOUD")

In[43]:

In [ ]:
import matplotlib.pyplot as plt
descending_order = df['Assignment group'].value_counts().sort_values(ascending=False).index
plt.subplots(figsize=(22,5))
 
ax=sns.countplot(x='Assignment group', data=df, color='royalblue',order=descending_order)
ax.set_xticklabels(ax.get_xticklabels(), rotation=45, ha="right")
plt.tight_layout()
plt.show()

#### Feature Generation using Bag of Words

Bag-of-words model(BoW ) is the simplest way of extracting features from the text. BoW converts text into the matrix of occurrence of words within a document. This model concerns about whether given words occurred or not in the document.

In[49]:

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer

In [ ]:
cv = CountVectorizer(lowercase=True,stop_words='english',ngram_range = (1,1))
text_counts= cv.fit_transform(df['Description'])

#### Split train and test set

In[50]:

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(
    text_counts, df['Description'], test_size=0.3, random_state=1)

#### Model Building and Evaluation

In[51]:

In [ ]:
from sklearn.naive_bayes import MultinomialNB
#Import scikit-learn metrics module for accuracy calculation
from sklearn import metrics
# Model Generation Using Multinomial Naive Bayes
clf = MultinomialNB().fit(X_train, y_train)
predicted= clf.predict(X_test)
print("MultinomialNB Accuracy:",metrics.accuracy_score(y_test, predicted))

#### Feature Generation using TF-IDF

#### TFIDF Vectorization<br>
- We now Use TFIDF Vectorization method to convert the Words in the description to number vectors.

In[52]:

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer
tf=TfidfVectorizer()
text_tf= tf.fit_transform(df['Description'])

#### Split train and test set (TF-IDF)

In[54]:

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(
    text_tf, df['Description'], test_size=0.3, random_state=123)

#### Model Building and Evaluation (TF-IDF)

In[55]:

In [ ]:
from sklearn.naive_bayes import MultinomialNB
from sklearn import metrics
# Model Generation Using Multinomial Naive Bayes
clf = MultinomialNB().fit(X_train, y_train)
predicted= clf.predict(X_test)
print("MultinomialNB Accuracy:",metrics.accuracy_score(y_test, predicted))

#### Handling Data Imbalance

In[ ]:

In[ ]:

In[ ]:

In [ ]:
sample = df.groupby(['Assignment group'])
regroup=[]
for grp in df['Assignment group'].unique():
    if(sample.get_group(grp).shape[0]<10):
        regroup.append(grp)
print('Found {} groups which have under 10 samples'.format(len(regroup)))
df['Assignment group']=df['Assignment group'].apply(lambda x : 'misc_grp' if x in regroup  else x)

Unique Groups check 

In [ ]:
df['Assignment group'].unique()

In[ ]:

In [ ]:
targetClassCnt=df['Assignment group'].value_counts()
targetClassCnt

In[ ]:

In [ ]:
df['Assignment group'].unique()

In[ ]:

In [126]:
df['Assignment group'].nunique()

74

In [125]:
deterministic_rules = {'grp_55' : {'finance_app'},
         'grp_6' : {'delivery'},
         'grp_55' : {'finance_app'},
         'grp_50' : {'virus','symantec'},
         'grp_45' : {'print','printed'},
         'grp_4' : {'cisco access point','switch'},
         'grp_3' : {'bad monitor'},
         'grp_36': {'timecards','time cards'},
         'grp_15': {'crm'},
         'grp_16' : {'collaboration_platform'},
         'grp_17' : {'reset passwords for'},
         'grp_62': {'security incidents'},
         'grp_8' : {'job hr_payroll_na', 'job SID_'},
         'grp_9' : {'failed in job_scheduler','abended job in job_scheduler'},
         'grp_68' : {'expense report'},
         'grp_0' : {'skype' , 'windows', 'account locked', 'vpn', 'password reset', 'account unlock', 'password', 'single sign', 
        'outlook', 'browser', 'login' ,'wi fi', 'wifi', 'logging' ,'wireless', 'phone','mail', 'mailbox',
        'unable to connect', 'log into' , 'erp SID_34 password reset' , 'change password', 'erp SID_', 'sign in',
        'audio issue', 'blank call'}
       }

In [117]:
deterministic_rules

{'grp_55': {'finance_app'},
 'grp_6': {'delivery'},
 'grp_50': {'symantec', 'virus'},
 'grp_45': {'print', 'printed'},
 'grp_4': {'cisco access point', 'switch'},
 'grp_3': {'bad monitor'},
 'grp_36': {'time cards', 'timecards'},
 'grp_15': {'crm'},
 'grp_16': {'collaboration_platform'},
 'grp_17': {'reset passwords for'},
 'grp_62': {'security incidents'},
 'grp_8': {'job SID_', 'job hr_payroll_na'},
 'grp_9': {'abended job in job_scheduler', 'failed in job_scheduler'},
 'grp_68': {'expense report'},
 'grp_0': {'account locked',
  'account unlock',
  'audio issue',
  'blank call',
  'browser',
  'change password',
  'erp SID_*',
  'erp SID_34 password reset',
  'log into',
  'logging',
  'login',
  'mail',
  'mailbox',
  'outlook',
  'password',
  'password reset',
  'phone',
  'sign in',
  'single sign',
  'skype',
  'unable to connect',
  'vpn',
  'wi fi',
  'wifi',
  'windows',
  'wireless'}}

In [119]:
def deterministicModelling(tokenList):
    for word in tokenList:
        for key, val in deterministic_rules.items():
            for i in val:
                if(i==word):
                    return key;
    return '';

In [124]:
grp_predicted=deterministicModelling(all_data_cleaned[1])
print()
if(grp_predicted==''):
    #Call the model
    print('calling the model')
else :
    print('The ticket should be assigned to grp :' , grp_predicted)


The ticket should be assigned to grp : grp_0
